Move "pylib" to "pythonx" for shared Python code

mhaig opened the pull request:
BREAKING CHANGE
Shared python code folders should now follow the "python2", "pythonx",
"python3" pattern described in ``:help python-special-path``

Update folder structure and ``vimrc`` to:
* Use ``python-special-path`` and only manually augment the Python path
  if the feature is not available (Vim < 7.3.1163)
* Update AugmentPythonPath() to use "python2" and "pythonx" instead of
  "pylib"

mhaig commented:
This one was fun and allowed me to learn a lot about Vim's Python integration and path handling!

I've been playing with NeoVim and with Vim in Windows and was experiencing issues using UltiSnips.  I found that the `AugmentPythonPath()` approach in vimrc does not update sys.path for Python 3 but UltiSnips will pick Python 3 over Python 2 if Python 3 is available.  When using Vim in Windows or using NeoVim, both version of Python were available, causing UltiSnips to pick Python 3 but then anything that required pylib was broken.

I spent a while trying different approaches to getting the path updated in both versions of Python before I stumbled on **python-special-path** and ultimately settled on that direction.  I left `AugmentPythonPath()` in-place but updated the child directories ("python2", "pythonx" vs "pylib") and only call it when Vim version < 7.3.1163 with Python support is detected.

I regression tested as best I could using versions of Vim around 7.3.1163.  I discovered that newer versions of Python 3 do not play well with older versions of Vim and ultimately decided to only call `AugmentPythonPath()` with Python 2 if Python 2 is available.

I realized I did not test this final solution in Windows.  I'll be sure to do that tomorrow and will reject if there are issues.

mhaig commented:
Works as expected in Windows!

drmikehenry commented:
Thanks for looking into this.  Python 2's life is drawing down; support
disappears January 1, 2020 (https://pythonclock.org), but it's not always
possible to ensure Python 3 is available everywhere we want to use Vim.  So we
need to make sure things as much as we can in cases where we've got only one
Python version available as well as when we've got both.

Spurred on by this pull request, I've looked a bit more into how Vim supports
Python, and looked through vimfiles to see how we're using it.  There are a
number of places where we explicitly invoke `:python`, and those won't work when
Vim is built without Python 2 support; they also fail on builds of Vim with both
versions supported if Python 3 is loaded before Python 2 (using ``:python3`` or
``has('python3')``).

Vim does support a `:pythonx` command that will use whichever version of Python
is already loaded; if none has been loaded yet, by default it will probe for
Python 3 first, then Python 2.  The probing order can be changed by setting
`pyxversion`.  I think all of our custom uses of Python can be made to work with
`:pythonx` so that they'll work with both versions of Python.  There are some
plugins that aren't built for that yet.  A search through vimfiles shows the
following:

    ag '\bpy(\b|th).*<<' 

    autoload/pythoncomplete.vim:90:python << PYTHONEOF
    bundle/fontdetect/autoload/fontdetect.vim:52:python << endpython
    bundle/ultisnips/doc/UltiSnips.txt:375:   Python << EOF
    bundle/localvimrc/plugin/localvimrc.vim:998:  let s:localvimrc_python_start = "pythonx << EOF"
    bundle/localvimrc/plugin/localvimrc.vim:1000:  let s:localvimrc_python_start = "python << EOF"
    bundle/localvimrc/plugin/localvimrc.vim:1002:  let s:localvimrc_python_start = "python3 << EOF"
    bundle/fireplace/autoload/nrepl/fireplace_connection.vim:131:python << EOF
    syntax_checkers/rst/rstsphinx.vim:16:python << endpython
    vimrc:482:python << endpython
    vimrc:3926:    python << endpython
    vimrc:5042:python << endpython

It looks like the localvimrc plugin will handle things fine.  No plugins seem to
be exclusively needing Python 3; but the uses in vimrc and the following plugins
seem likely to break with Python 3 only:
- pythoncomplete
- fontdetect
- fireplace
- rstsphinx.vim

It seems like we ought to make sure that all of vimfiles will work with Python
3.  There may be other uses that the above search doesn't uncover; I guess we'll
find them when we discover they aren't working someday.

It seems like we ought to try to use `:pythonx` (and the `pythonx/` directory)
in vimfiles as much as possible, and to use Python 2/3 straddling code in
general until we no longer need to support builds of Vim that have only Python
2.  It also seems reasonable to attempt to use Python 3 first, then fall back to
Python 2.  This happens naturally when using `has('pythonx')` with the
default setting for `pyxversion`.

Also, it makes sense to use Vim's "python-special-path" feature when it's
available as done in your pull request; when that feature isn't available, we
should fall back to augmenting `sys.path` manually.  I think we'd want to
augment for Python 3 as well, unless you've discovered it doesn't work somehow
(i.e., we should add `python2/` directories when `has('python')` is true, and
`python3/` directories otherwise).  I think this could be done via an initial
check for `has('pythonx')`, and if that's true, we have the `:pythonx` command
available to use everywhere; in addition, exactly one of `has('python')` and
`has('python3')` will be true, allowing us to decide which directories to add to
`sys.path`.

Does this overall plan make sense?  Maybe @jszakmeister knows if the fireplace
plugin has an upgrade with Python 3 support (or else thinks we don't need it to
work with Python 3), and whether rstsphinx.vim can be updated for Python 3.  I
should update my fontdetect plugin for Python 3.  pythoncomplete.vim doesn't
seem to have any updates; the version in
https://www.vim.org/scripts/script.php?script_id=1542 (which is embedded in
vimfiles) is the same as the version distributed in Vim 8.1.  Maybe we need to
search out better completion support for Python.  How does this sound overall?

mhaig commented:
I came to this same conclusion yesterday as I realized when I build Vim for
Linux I was always ending up with `:python` commands  working and not `:python3`
commands.  I'll admit that when I submitted this pull request I was focused on
UltiSnip as it failed loudly while the other plugins and vimrc fail silently.

When I started the work I mistakenly assumed that `:python` and `has('python')`
would "just work" if either version of Python was installed and
`:python3`/`has('python3')` was for Python 3 specifically.  I discovered the
flaw in my logic while working this but never thought to go back to ensure
complete Python 3 support, again focusing on UltiSnip.  Turns out I had just
imagined `:pythonx` before I knew it existed!

I find the approach outlined above to make complete sense and something I
started undertaking before this comment.  `:pythonx` and `pyxversion` were not
introduced until 8.0.0251 so there will need to be some extra logic in `vimrc`
but nothing impossible, something along the lines of:
```
if has('python3')
endif
```
is enough to force older versions of Vim to use the Python 3 interpreter if it's
available.

Regarding augmenting `sys.path` with `python3/` directories.  This would be
simple enough and I had it in my branch at one point.  The problem I found was
testing.  In my regression testing I found there must be enough changes in
Python 3 since Vim 7.3.1163 that I could never get all the features in the `vim`
module to work with Python 3.5 (oldest Python 3 interpreter I had available at
the time).  Once I confirmed that the original `AugmentPythonPath()` only ran
for Python 2 I assumed that Python 2 support with older versions of Vim was
sufficient and dropped handling of Python 3.

I like the completeness of the approach you detailed above and agree.  My
efforts forward will be updating all of vimfiles to support Python 3,
configuring vimrc to default to Python 3 if available, and proper regression
testing.

drmikehenry commented:
Inspired by the way localvimrc.vim detects the Python version in one place and
uses `:execute` to avoid spreading that logic around, I think we can setup a
global variable `g:Python` to use in a similar way, defining it in the "Python
path management" section.  That's about as early as we can make it while still
allowing a user to override `:pyxversion` or do something else that influences
which version of Python is active.  Below is what I have in mind for that aspect
(minus the additional fancies needed for `AugmentPythonPath()` specifically):

    " -------------------------------------------------------------
    " Python path management
    " -------------------------------------------------------------

    if has('pythonx')
        let g:Python = 'pythonx'
    elseif has('python3')
        let g:Python = 'python3'
    elseif has('python')
        let g:Python = 'python'
    else
        let g:Python = ''
    endif

    if g:Python != ''
    function! AugmentPythonPath()
        execute g:Python . ' << endpython'

    import vim
    ...
    endpython
    endfunction

I think this will work across Vim versions and allow the rest of vimfiles ignore
the Python version issue.

So this sounds like a good plan to me; thanks for getting this rolling!

drmikehenry commented:
We're partway there.  The following portions of vimfiles have been addressed; they are either working with Python 2 and Python3 or have been removed:
- fireplace (removed)
- fontdetect
- rstsphinx.vim
- tool/markdown2ctags.py
- tool/execargs.py
- vimrc `python` invocations
- Python indentation support (updated)
- python syntax bundle (removed; was unmaintained)
- buildtool

Still remaining are:
- pythoncomplete plugin (unmaintained; perhaps replace this with jedi? https://github.com/davidhalter/jedi-vim)
- "Python Path" section of notes.txt
- `pylib/` directories
- `UltiSnips/*.py` files

mhaig commented:
Hmm, I did not intend that to happen, I was just trying to re-align my branch!

At this point is it just better to close this request and make a new one once I have everything addressed?

drmikehenry commented:
Feel free to create a new pull request if that makes things easier, or you can try rebasing your branch.

mhaig commented:
I got my branch cleaned up but I still need to test, double check `UltiSnips/*.py`, and add another thought to notes.txt.

I'll let you know when it's ready for more eyes.

drmikehenry commented:
Agreed regarding the bias toward python2; we don't want the python3 directories to feel unloved :-)  Thanks for the progress.

mhaig commented:
Wasn't as much as I thought left to do!

I'll get more time on it through the holidays but I don't foresee any major changes unless there are issues.

Thanks for the patience!

mhaig commented:
OK take a look at this, I think I captured everything.  It's a force push to try and keep things clean and make sure the commit message matches the final solution.

I pushed a branch "augment_python_path_backup" that has more explicit commits if you want to see a better history of the changes.

drmikehenry commented:
Sorry for the delay; I've merged your recent changes to support Python 3 within vimfiles.  Thanks for your persistence in getting this support in place :-)

jszakmeister commented:
Thank you @mhaig!
